tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs

```diff
@@ -0,0 +1,184 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Azure.Sdk.Tools.CodeOwnersParser
+{
+    /// <summary>
+    /// Represents a CODEOWNERS file entry that matched to targetPath from
+    /// the list of entries, assumed to have been parsed from CODEOWNERS file.
+    ///
+    /// To obtain the value of the matched entry, reference "Value" member.
+    /// </summary>
+    internal class MatchedCodeOwnerEntry
+    {
+        public readonly CodeOwnerEntry Value;
+
+        private static readonly char[] unsupportedChars = { '[', ']', '!', '?' };
+
+        public MatchedCodeOwnerEntry(List<CodeOwnerEntry> entries, string targetPath)
+        {
+            this.Value = FindOwnersForClosestMatch(entries, targetPath);
+        }
+
+        /// <summary>
+        /// Returns a CodeOwnerEntry from codeOwnerEntries that matches targetPath
+        /// per algorithm described in:
+        /// https://git-scm.com/docs/gitignore#_pattern_format
+        /// and
+        /// https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-syntax
+        ///
+        /// If there is no match, returns "new CodeOwnerEntry()".
+        /// </summary>
+        private static CodeOwnerEntry FindOwnersForClosestMatch(
+            List<CodeOwnerEntry> codeownersEntries,
+            string targetPath)
+        {
+            // targetPath is assumed to be absolute w.r.t. repository root, hence we ensure
+            // it starts with "/" to denote that.
+            if (!targetPath.StartsWith("/"))
+                targetPath = "/" + targetPath;
+
+            // We do not trim or add the slash ("/") at the end of the targetPath because its
+            // presence influences the matching algorithm:
+            // Slash at the end denotes the target path is a directory, not a file, so it might
+            // match against a CODEOWNERS entry that matches only directories and not files.
+
+            // Entries below take precedence, hence we read the file from the bottom up.
+            // By convention, entries in CODEOWNERS should be sorted top-down in the order of:
+            // - 'RepoPath',
+            // - 'ServicePath'
+            // - and then 'PackagePath'.
+            // However, due to lack of validation, as of 12/29/2022 this is not always the case.
+            for (int i = codeownersEntries.Count - 1; i >= 0; i--)
+            {
+                string codeownersPath = codeownersEntries[i].PathExpression;
+                if (ContainsUnsupportedCharacters(codeownersPath))
```

Comment:
From a diagnostic standpoint it is probably interesting to know we encountered an unsupported char.

Reply:
I'll add diagnostic output in the follow-up PR:

- #5063